### DISCUSSION TITLE
[Internal] Move away from npm and lerna bootstrap by using yarn and yarn workspaces

### DISCUSSION THREAD

**Discussion by hieuxlu** [webdriverio/webdriverio#6193] [Technical Discussion]
@christian-bromann This issue is created in order to solve the problem in https://github.com/webdriverio/webdriverio/pull/6085#discussion_r534558596

## Problems
When we start providing Typescript declaration autogenerated by tsc and provide it in package.json to avoid manually providing main `d.ts` typings, we'll encounter `Conflicting definitions for 'node'` error as mentioned in https://github.com/webdriverio/webdriverio/pull/6085#issuecomment-735399213

For example, if we change
```diff
+  "types": "./build/index.d.ts",
-  "types": "./allure-reporter.d.ts"
```

We'll have the below error when running `test:typings`
```
error TS4090: 9 TS4090: Conflicting definitions for 'node' found at 'D:/repos/webdriverio/packages/webdriver/node_modules/@types/node/index.d.ts' and 'D:/repos/webdriv
      Conflicting definitions for 'node' found at 'D:/repos/webdriverio/packages/webdriver/node_modules/@types/node/index.d.ts' 
```

## Solutions
I believe this was caused by npm and lerna not having any feature to fix a resolution of `@types/node` package. Yarn does have such feature by using [resolutions property in package.json](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/). Yarn also supports monorepo with [workspaces](https://classic.yarnpkg.com/en/docs/workspaces/), and [supported by lerna](https://github.com/lerna/lerna/blob/main/commands/bootstrap/README.md#--use-workspaces) by providing `npmClient: yarn` and `useWorkspaces: true`

IMHO, moving aways from npm and lerna bootstrap, and use yarn & yarn workspaces together with lerna is the only solution if we want to ship auto generated Typescript typings, and don't have to manually provide `.d.ts` typings in the future. 

So I suggest we discuss about this proposal. I also created a POC, which shows it is trivial to move to Yarn. I face only one minor problem in unit tests with `puppeteer` trying to resolve `debug` module via `fs`, but `fs` is mocked in Jest. https://github.com/hieuxlu/webdriverio/commit/c149647f08f93841600190d606f17460cac5d9c0

**Comment by christian-bromann** [comment]
I have never experienced the issue of having conflicting definitions for node there for I don't see a reason to switch to Yarn.

**Comment by hieuxlu** [comment]
@christian-bromann As I said, it's a time-bomb. To reproduce, just reference node typing in `allure-reporter.d.ts` as below
```diff
+ /// <reference types="node" />

declare namespace AllureReporter {
```

Then execute `npm run test:typings`, you will see that tests for `wdio-sync` will fails.
Maybe the cause is because of the way we're naively copying packages to node_modules folder for typing tests, which also copy all dependencies. In this case, I suspect `@wdio/allure-reporter` is the first package that references `@wdio/reporter`, so when both reference `node`, this issue will happen. 

So I guess either we need to change the way we test typings, or we'll need to workaround by not using auto-generated typings for `@wdio/allure-reporter`. 

If both are not possible, I suggest to move to Yarn as our last resort. Because the root cause is lerna is cloning different copies of `@types/node` everywhere in our repos in a way that we can't control. Yarn resolution will fix this.

**Comment by mgrybyk** [comment]
Let's solve the problem at the moment they appear.

**Comment by hieuxlu** [comment]
@mgrybyk It's happening right now and is blcoking #6085

**Comment by dylanlive** [comment]
I'm currently coming across this while setting up the repo for the first time.
```
> webdriverio-monorepo@ ts:sync ~/Documents/open_source/dylan/webdriverio
> cd tests/typings/sync && npx tsc

../../../node_modules/ts-node/dist/repl.d.ts:1:23 - error TS4090: Conflicting definitions for 'node' found at '~/Documents/open_source/dylan/webdriverio/node_modules/@types/node/index.d.ts' and '~/Documents/open_source/dylan/webdriverio/packages/webdriverio/node_modules/@types/node/index.d.ts'. Consider installing a specific version of this library to resolve the conflict.

1 /// <reference types="node" />
```

I ran through the Set Up Project Steps, cloning, `npm install`, `npm run setup-full` and then ran `npm run test` and came across this error.

Running node `v12.21.0`

I'm unsure how to proceed

**Reply by christian-bromann** [reply]
@dylanlive Hey Dylan, just merged a patch that should fix this bug.
